storage/gcp/gcp.go

```diff
@@ -114,47 +113,63 @@ func New(ctx context.Context, cfg Config) (*Storage, error) {
 	}
-	return ""
-}
-
-// The location to which the tile is written is defined by the tile layout spec.
-func (s *Storage) setTile(ctx context.Context, level, index, logSize uint64, tile [][]byte) error {
-	t := slices.Concat(tile...)
 	tPath := layout.TilePath(level, index, logSize)
-	klog.V(2).Infof("StoreTile: %s", tPath)
+	klog.V(2).Infof("StoreTile: %s (%d entries)", tPath, len(tile.Nodes))
 
-	return s.objStore.setObject(ctx, tPath, t, &gcs.Conditions{DoesNotExist: true})
+	return s.objStore.setObject(ctx, tPath, data, &gcs.Conditions{DoesNotExist: true})
 }
 
-// getTile returns the tile at the given tile-level and tile-index for the specified log size, or
-// an error if it doesn't exist.
-func (s *Storage) getTile(ctx context.Context, level, index, logSize uint64) ([][]byte, error) {
+// getTile returns the tile at the given tile-level and tile-index for the specified log size.
+//
+// Returns a wrapped os.ErrNotExist if the tile does not exist.
+func (s *Storage) getTile(ctx context.Context, level, index, logSize uint64) (*api.HashTile, error) {
 	objName := layout.TilePath(level, index, logSize)
 	data, _, err := s.objStore.getObject(ctx, objName)
 	if err != nil {
 		if errors.Is(err, gcs.ErrObjectNotExist) {
 			// Return the generic NotExist error so that higher levels can differentiate
 			// between this and other errors.
-			return nil, os.ErrNotExist
+			return nil, fmt.Errorf("%v: %w", objName, os.ErrNotExist)
 		}
 		return nil, err
 	}
+	t := &api.HashTile{}
+	return t, t.UnmarshalText(data)
+}
 
-	// Recreate the tile slices
-	l := len(data)
-	if m := l % sha256.Size; m != 0 {
-		return nil, fmt.Errorf("invalid tile data size %d", l)
+// getEntryBundle returns the entry bundle at the location implied by the given index and treeSize.
+//
+// Returns a wrapped os.ErrNotExist if the bundle does not exist.
+func (s *Storage) getEntryBundle(ctx context.Context, bundleIndex uint64, logSize uint64) (*api.EntryBundle, error) {
+	objName := layout.EntriesPath(bundleIndex, logSize)
+	data, _, err := s.objStore.getObject(ctx, objName)
+	if err != nil {
+		if errors.Is(err, gcs.ErrObjectNotExist) {
+			// Return the generic NotExist error so that higher levels can differentiate
+			// between this and other errors.
+			return nil, fmt.Errorf("%v: %w", objName, os.ErrNotExist)
+		}
+		return nil, err
 	}
-	t := make([][]byte, l/sha256.Size)
-	for i := range t {
-		t[i] = data[i*sha256.Size : (i+1)*sha256.Size]
+
+	r := &api.EntryBundle{}
+	return r, r.UnmarshalText(data)
+}
+
+// setEntryBundle idempotently stores the entry bundle at the location implied by the bundleIndex and treeSize.
+func (s *Storage) setEntryBundle(ctx context.Context, bundleIndex uint64, logSize uint64, bundle *api.EntryBundle) error {
+	objName := layout.EntriesPath(bundleIndex, logSize)
+	data, err := bundle.MarshalText()
+	if err != nil {
+		return err
 	}
-	return t, nil
+	return s.objStore.setObject(ctx, objName, data, &gcs.Conditions{DoesNotExist: true})
```

Comment:
Can you put a comment here to explain that retries will work if the content of the file is the same